# Kullancağımız kütüphaneleri import ediyoruz.

In [189]:
import numpy as np
import pandas as pd
import sklearn

# Verisetimizi okuyoruz.

In [190]:
df = pd.read_csv('Placement_Data_Full_Class.csv')
df.head()

sl_no gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p  \
0      1      M  67.00   Others  91.00   Others  Commerce     58.00   
1      2      M  79.33  Central  78.33   Others   Science     77.48   
2      3      M  65.00  Central  68.00  Central      Arts     64.00   
3      4      M  56.00  Central  52.00  Central   Science     52.00   
4      5      M  85.80  Central  73.60  Central  Commerce     73.30   

    degree_t workex  etest_p specialisation  mba_p      status    salary  
0   Sci&Tech     No     55.0         Mkt&HR  58.80      Placed  270000.0  
1   Sci&Tech    Yes     86.5        Mkt&Fin  66.28      Placed  200000.0  
2  Comm&Mgmt     No     75.0        Mkt&Fin  57.80      Placed  250000.0  
3   Sci&Tech     No     66.0         Mkt&HR  59.43  Not Placed       NaN  
4  Comm&Mgmt     No     96.8        Mkt&Fin  55.50      Placed  425000.0

# Modelimizde kullanmayacağımız kolonları datasetimizden düşürüyoruz.

In [191]:
df.drop(['sl_no','ssc_b','hsc_b','etest_p','salary'], axis=1, inplace=True)
df.head()

gender  ssc_p  hsc_p     hsc_s  degree_p   degree_t workex specialisation  \
0      M  67.00  91.00  Commerce     58.00   Sci&Tech     No         Mkt&HR   
1      M  79.33  78.33   Science     77.48   Sci&Tech    Yes        Mkt&Fin   
2      M  65.00  68.00      Arts     64.00  Comm&Mgmt     No        Mkt&Fin   
3      M  56.00  52.00   Science     52.00   Sci&Tech     No         Mkt&HR   
4      M  85.80  73.60  Commerce     73.30  Comm&Mgmt     No        Mkt&Fin   

   mba_p      status  
0  58.80      Placed  
1  66.28      Placed  
2  57.80      Placed  
3  59.43  Not Placed  
4  55.50      Placed

# Kategorik ve numerik kolonları listeleyip kategorik, numerik ve target dataframeleri oluşturuyoruz.

In [192]:
col_num = ['ssc_p','hsc_p','degree_p','mba_p'] #liste

col_cat = ['gender','hsc_s','degree_t','specialisation'] #liste

col_tar = 'status' #string

df_num = df[col_num]
df_cat = df[col_cat]
y = df[col_tar]

# Kategorik verilere one-hot encoding uyguluyoruz.# 

In [193]:
df2 = df_cat
df2.head()

gender     hsc_s   degree_t specialisation
0      M  Commerce   Sci&Tech         Mkt&HR
1      M   Science   Sci&Tech        Mkt&Fin
2      M      Arts  Comm&Mgmt        Mkt&Fin
3      M   Science   Sci&Tech         Mkt&HR
4      M  Commerce  Comm&Mgmt        Mkt&Fin

In [194]:
sg_dum = pd.get_dummies(df2)
sg_dum.head()

gender_F  gender_M  hsc_s_Arts  hsc_s_Commerce  hsc_s_Science  \
0         0         1           0               1              0   
1         0         1           0               0              1   
2         0         1           1               0              0   
3         0         1           0               0              1   
4         0         1           0               1              0   

   degree_t_Comm&Mgmt  degree_t_Others  degree_t_Sci&Tech  \
0                   0                0                  1   
1                   0                0                  1   
2                   1                0                  0   
3                   0                0                  1   
4                   1                0                  0   

   specialisation_Mkt&Fin  specialisation_Mkt&HR  
0                       0                      1  
1                       1                      0  
2                       1                      0  
3                       0                      1  
4                       1                      0

# One-hot encoding uyguladığımız verilerimiz ile numerik verilerimizi birleştiriyoruz.

In [195]:
X = pd.concat([df_num,sg_dum],axis=1)
print(X.isna().sum().sum()) # 0 olmalıdır.
X.head()

0


ssc_p  hsc_p  degree_p  mba_p  gender_F  gender_M  hsc_s_Arts  \
0  67.00  91.00     58.00  58.80         0         1           0   
1  79.33  78.33     77.48  66.28         0         1           0   
2  65.00  68.00     64.00  57.80         0         1           1   
3  56.00  52.00     52.00  59.43         0         1           0   
4  85.80  73.60     73.30  55.50         0         1           0   

   hsc_s_Commerce  hsc_s_Science  degree_t_Comm&Mgmt  degree_t_Others  \
0               1              0                   0                0   
1               0              1                   0                0   
2               0              0                   1                0   
3               0              1                   0                0   
4               1              0                   1                0   

   degree_t_Sci&Tech  specialisation_Mkt&Fin  specialisation_Mkt&HR  
0                  1                       0                      1  
1                  1                       1                      0  
2                  0                       1                      0  
3                  1                       0                      1  
4                  0                       1                      0

# Numerik verilerimize normalizasyon uyguluyoruz.

In [196]:
from sklearn.preprocessing import Normalizer
datafr = Normalizer().fit_transform(X)
datafr

array([[0.47862711, 0.65007563, 0.41433392, ..., 0.00714369, 0.        ,
        0.00714369],
       [0.52504165, 0.5184232 , 0.51279752, ..., 0.00661845, 0.00661845,
        0.        ],
       [0.50927804, 0.53278318, 0.50144299, ..., 0.        , 0.00783505,
        0.        ],
       ...,
       [0.48388535, 0.48388535, 0.52721836, ..., 0.        , 0.00722217,
        0.        ],
       [0.5704531 , 0.5087825 , 0.44711189, ..., 0.        , 0.        ,
        0.00770883],
       [0.53072062, 0.49648058, 0.45368053, ..., 0.        , 0.        ,
        0.00856001]])

In [197]:
X = pd.DataFrame(datafr,columns=['ssc_p', 'hsc_p', 'degree_p', 'mba_p', 'gender_F', 'gender_M',
       'hsc_s_Arts', 'hsc_s_Commerce', 'hsc_s_Science', 'degree_t_Comm&Mgmt',
       'degree_t_Others', 'degree_t_Sci&Tech', 'specialisation_Mkt&Fin',
       'specialisation_Mkt&HR'])
X.head()

ssc_p     hsc_p  degree_p     mba_p  gender_F  gender_M  hsc_s_Arts  \
0  0.478627  0.650076  0.414334  0.420049       0.0  0.007144    0.000000   
1  0.525042  0.518423  0.512798  0.438671       0.0  0.006618    0.000000   
2  0.509278  0.532783  0.501443  0.452866       0.0  0.007835    0.007835   
3  0.509514  0.473120  0.473120  0.540721       0.0  0.009098    0.000000   
4  0.588781  0.505061  0.503003  0.380855       0.0  0.006862    0.000000   

   hsc_s_Commerce  hsc_s_Science  degree_t_Comm&Mgmt  degree_t_Others  \
0        0.007144       0.000000            0.000000              0.0   
1        0.000000       0.006618            0.000000              0.0   
2        0.000000       0.000000            0.007835              0.0   
3        0.000000       0.009098            0.000000              0.0   
4        0.006862       0.000000            0.006862              0.0   

   degree_t_Sci&Tech  specialisation_Mkt&Fin  specialisation_Mkt&HR  
0           0.007144                0.000000               0.007144  
1           0.006618                0.006618               0.000000  
2           0.000000                0.007835               0.000000  
3           0.009098                0.000000               0.009098  
4           0.000000                0.006862               0.000000

# Gerekli kütüphanelerimizi import edip train ve test datalarını oluşturuyoruz.

In [149]:
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.35,random_state=10)

# Farklı ML algoritmaları ile modelimizi eğitip doğruluk oranlarına bakıyoruz.

In [151]:
from sklearn import tree

clf = tree.DecisionTreeClassifier()
clf = clf.fit(X_train, y_train)

y_pred = clf.predict(X_test)

accuracy = accuracy_score(y_test,y_pred)
accuracy*100

89.47368421052632

In [152]:
from sklearn.naive_bayes import GaussianNB

clf2 = GaussianNB()
clf2.fit(X_train,y_train)

y_pred = clf2.predict(X_test)

accuracy = accuracy_score(y_test,y_pred)
accuracy*100

76.31578947368422

In [153]:
from sklearn.neighbors import KNeighborsClassifier

clf3 = KNeighborsClassifier(n_neighbors=7)
clf3.fit(X_train,y_train)

y_pred = clf3.predict(X_test)

accuracy = accuracy_score(y_test,y_pred)
accuracy*100

84.21052631578947

In [154]:
from sklearn.svm import SVC

clf4 = SVC(kernel='rbf',probability=True)
clf4.fit(X_train,y_train)

y_pred = clf4.predict(X_test)

accuracy = accuracy_score(y_test,y_pred)
accuracy*100

67.10526315789474

In [184]:
from sklearn.ensemble import RandomForestClassifier

clf5 = RandomForestClassifier(max_depth=10)
clf5.fit(X_train,y_train)

y_pred = clf5.predict(X_test)

accuracy = accuracy_score(y_test,y_pred)
accuracy*100

85.52631578947368

# Sonuç:
# En yüksek doğruluk oranını Decision Tree algoritmasında aldık.
# Modelimiz, istenilen bilgileri giren kullanıcının işe alınıp alınmayacağını %89.47 doğruluk oranı ile tahmin ediyor.
